include/revng-c/TypeNames/PTMLCTypeBuilder.h

```diff
@@ -0,0 +1,245 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng-c/Backend/DecompiledCCodeIndentation.h"
+#include "revng-c/Support/PTMLC.h"
+#include "revng-c/TypeNames/DependencyGraph.h"
+
+namespace ptml {
+
+class PTMLCTypeBuilder : public PTMLCBuilder {
+public:
+  using OutStream = ptml::PTMLIndentedOstream;
+
+protected:
+  /// The stream to print the result to.
+  std::unique_ptr<OutStream> Out;
+
+public:
+  /// When set to true, the types that are only ever depended on by a single
+  /// other type (for example, a nested struct), are printed inside the parent
+  /// type definition instead of separately.
+  bool EnableTypeInlining = true;
+
+  /// When set to true, function stack frame types are printed inside the struct
+  /// body as opposed to the type header.
+  bool EnableStackFrameInlining = true;
+
+  /// Because we are emitting older C, we cannot specify underlying enum type,
+  /// which means this that when we need to preserve enum size across
+  /// the compilation boundary (for example, when editing a type), we need
+  /// to resolve to a trick: we print the maximum value possible - and then use
+  /// it to figure out the original size.
+  /// Setting this flag to true enables printing of such a value.
+  bool EnablePrintingOfTheMaximumEnumValue = false;
+
+  /// All of our structs are packed by definition. Which means every single byte
+  /// has to occupied. In explicit padding mode, the padding is printed as
+  /// arrays of bytes. In implicit mode, a `_START_AT` annotation is printed
+  /// instead.
+  bool EnableExplicitPaddingMode = true;
+
+  /// When this flag is set to true, a special `_SIZE($bytes)` annotation is
+  /// printed with every struct definition.
+  bool EnableStructSizeAnnotation = false;
+
+  /// Sometimes you don't want to print everything. This lets you specify
+  /// a set of types that will be ignored by \ref typeDefinitions.
+  std::set<model::TypeDefinition::Key> TypesToOmit = {};
+
+  /// This the cache containing the names of artificial types (like array
+  /// wrappers).
+  std::map<model::UpcastableType, std::string> ArtificialNameCache = {};
+
+public:
+  static constexpr std::string_view structPaddingPrefix() {
+    return "_padding_at_";
+  }
+
+  static constexpr std::string_view artificialReturnValuePrefix() {
+    return "_artificial_struct_returned_by_";
+  }
+  static constexpr std::string_view artificialArrayWrapperPrefix() {
+    return "_artificial_wrapper_";
+  }
+
+  static constexpr std::string_view artificialReturnValueFieldPrefix() {
+    return "field_";
+  }
+  static constexpr std::string_view artificialArrayWrapperFieldName() {
+    return "the_array";
+  }
+
+private:
+  std::optional<DependencyGraph> DependencyCache = std::nullopt;
+  std::set<model::TypeDefinition::Key> TypesToInlineCache = {};
+  std::set<model::TypeDefinition::Key> StackFrameTypeCache = {};
```

Comment:
This one, I disagree on. These are cache, not configuration - external world shouldn't really even know about their existence.

For some context, in the new version, the options (as in `bool`s responsible for enabling these) are separate - you can enable one but not the other.

The caches, on the other hand, are not - `TypesToInlineCache` contains _all_ the types. While `StackFrameTypeCache` only shows which of them are stack frames.

There is no deeper reason for this that just that it made implementation easier.

Reply:
I like some of the changes you've done, like separating configurations for type inlining (in other types) and stack frame inlining (in function bodies).

I'll review all of it once again and come back to this.

Reply:
In any case the caches aren't caches.

My understanding of a cache is something that makes things faster but isn't supposed to alter behavior if something is missing in the cache. The only case where it might alter behavior is when some entry of the cache was supposed to be invalidated but it didn't, so you have a false hit with data that might be stale (hence wrong).

Instead I see:
```c++
  bool shouldInline(model::TypeDefinition::Key Key) const {
    if (not TypesToInlineCache.contains(Key)) {
      // This type is not allowed be inlined.
      return false;
    }
```

If the cache is empty this method says that everything should be inlined, which is wrong.

So, something of what you say about caches is not really adding up.

Reply:
In my mind, they are caches in a sense that there's nothing preventing you from computing this list every time you need it (since the rules for selecting such types are pretty straight forward). All they change is that they allow us to only do it once and act like it never changes (which is true for any pipe anyway).

Reply:
What I'm trying to say overall is that there is some logic that is external to the builder that has to decide what types (and what stack) can be inlined and what can't. And once that's decided, that info doesn't ever change for the whole lifetime of the builder. So it can effectively be a "configuration" of the builder, or a constructor parameter

Reply:
That being said, I'm not to attached. If you can think of a better name (or a better system for that matter), I'm all ears!

Reply:
> What I'm trying to say overall is that there is some logic that is external to the builder

ah, I see the misunderstanding now. The logic is _inside_ the builder now. Hence the cache thing.

Reply:
As for should inline, I also see the problem now. It was private initially, hence why I didn't bother recomputing the cache there. But then a use case forced me to make it public. My bad on 
not updating with regards to that.

Reply:
Wow, your intuition really did its thing here! Turns out there was indeed a case where `shouldInline` was invoked without first filling the map in!

Reply:
Done (I added an assert, since it don't think it's viable to make `shouldInline` depend on the model).